main/src/com/google/refine/expr/functions/Cross.java

```diff
@@ -41,24 +41,37 @@
 import com.google.refine.expr.WrappedCell;
 import com.google.refine.grel.ControlFunctionRegistry;
 import com.google.refine.grel.Function;
+import com.google.refine.model.Project;
 import com.google.refine.util.GetProjectIDException;
 import com.google.refine.util.LookupException;
 
 public class Cross implements Function {
 
+    public static final String INDEX_COLUMN_NAME = "_OpenRefine_Index_Column_Name_";
+
     @Override
     public Object call(Properties bindings, Object[] args) {
-        if (args.length == 3) {
+        if (args.length == 1 || args.length == 2 || args.length == 3) {
             // 1st argument can take either value or cell(for backward compatibility)
             Object v = args[0];
-            Object targetProjectName = args[1];
-            Object targetColumnName = args[2];
+            // if 2nd argument is omitted or set to "", use the current project name
+            Object targetProjectName = "";
+            boolean isCurrentProject = false;
+            if (args.length < 2 || args[1].equals("")) {
+                isCurrentProject = true;
+            } else {
+                targetProjectName = args[1];
+            }
+            // if 3rd argument is omitted or set to "", use the index column
+            Object targetColumnName = args.length < 3 || args[2].equals("") ? INDEX_COLUMN_NAME: args[2];
```

Comment:
Hmm...  what are we missing here?

I perform a `cross()` that has valid syntax and 3 args.
But there's no try wrapping here...so... silent error?
My expectation was that of some kind of error GIVEN the below scenario:

![image](https://user-images.githubusercontent.com/986438/79641614-469fad00-815e-11ea-83ff-f7ecd9ed820b.png)

Reply:
I don’t think it’s an error.

Indexes being numerical, you got to match them with numbers.

Regards, A.

Reply:
Sorry @antoine2711 I don't quite understand what you mean.  Please explain what I am asking GREL for with that syntax in my image and given the underlying data grid.

Reply:
> Sorry @antoine2711 I don't quite understand what you mean. Please explain what I am asking GREL for with that syntax in my image and given the underlying data grid.

You are doing `cross(cell.value, "CurrentProject", Use the internal index)`.

Since internal index is ALWAYS a number, cell.value MUST be a number to match & cross. In your example, 111 or 222 « could be » numbers, but since the preview doesn't use the green coloring to confirm it, can't know for sure, but at the look of the grey table in the background, neither 111 or 222 or 333 seams to be cast as number, but rather like strings, which will never match an index.

But, even if 111 or 222 where numbers, you would need 112 or 223 rows in your project to have a match.

Is it clearer? Regards, A.

Reply:
they are Strings (not numbers)

Reply:
> they are Strings (not numbers)

K. So, unless they are numerical and between 0 and « Max(row.index) », you can always expect `null` result value, and not an error.

Regards, A.

Reply:
No, it's much more confusing now.  I need a test case (a fuller test case) that shows how the 3rd argument works or doesn't.  I'd expect that single test case to show a single cross operation that is using the 3rd argument working against a test data grid of 3 rows x 2 columns.

I was also confused by the issue #2504 and still am, so that's part of the problem....me actually seeing some usefulness...which I have not yet, so I need everyone to help me here.

Reply:
@antoine2711 Would you mind helping everyone understand how the new `cross()` is supposed to work on a new wiki page or some Google document or Hackmd.io so that the community can leave comments?  Call it "Proposed Cross() in Issue #2504"  The issue #2504 is not clear to me, sorry.  Thanks.

Reply:
> No, it's much more confusing now. 
Your confusion seams to come from GREL cross function should support lookup by numbers #2461, which is committed into the master branch since almost a month.

Now, when we use the `cross()` function, matches can be done even if the datatype is not a string. 

> I need a test case (a fuller test case) that shows how the 3rd argument works or doesn't. I'd expect that single test case to show a single cross operation that is using the 3rd argument working against a test data grid of 3 rows x 2 columns.

I think you are raising a good point, that #2461 could have more tests.
What should we add to this file? `main/tests/server/src/com/google/refine/expr/functions/CrossTests.java`

> I was also confused by the issue #2504 and still am, so that's part of the problem....me actually seeing some usefulness...which I have not yet, so I need everyone to help me here.
The real added value to Issue #2504 is to be able to do cross() with the internal index of rows, that is not a column, but can easily behave like one. It is very powerful as it actually enables every project to be used like an index array. I basically makes it possible to access every cells of every projects. It's very flexible.

Should we discuss of datatypes with the cross() function in Issue #2461 or here? It much more relevent there.

Regards, Antoine

Reply:
> @antoine2711 Would you mind helping everyone understand how the new `cross()` is supposed to work on some Google document or Hackmd.io so that the community can leave comments? Call it "Proposed Cross() in Issue #2504" The issue #2504 is not clear to me, sorry. Thanks.

Yes, but like I wrote, it would be about Issue #2461. The situation you have here is the same you would have with an added column called ID that would contain numeric values of index (`row.index`).

This is the first thing I was planning to do on the new documentation system. And it's important to have that ready BEFORE v3.4 goes out, since it already committed in the master branch. I could go in the Wiki, and start a new page, that could be exposed once v3.4 is out.

Regards, 
   Antoine

Reply:
The community can't leave comments on a GitHub Wiki page, unfortunately.  Please use a collaborative environment.  My suggestion is https://hackmd.io

Reply:
Hi, @thadguidry , with #2571, we are now using string match for the cross() function now. So the result of your test above will be:

![test1](https://user-images.githubusercontent.com/29347603/81515718-f637fb80-9367-11ea-8fc2-1532066bed41.png)

Here the index range is [0, 9], so in column "#@!c", only the value "3" got a match.

But there is another problem here: 

Both the embedded GREL variable "rowIndex" and the index used by the cross() function start from 0, so "cross(rowIndex)" will match the same column, which is what we expected. 

![test3](https://user-images.githubusercontent.com/29347603/81517080-55980a80-936c-11ea-9c3a-82aab3fef458.png)


However the row index in the UI starts from 1. This is confusing, because "cross(3)" actually match the 4th row.

![test2](https://user-images.githubusercontent.com/29347603/81516477-98f17980-936a-11ea-857e-c8b3a34ad65c.png)

My suggestion is that we change the "rowIndex" and the index used by the cross() function both to start from 1, so they will be consistent with the UI index. But I am not sure if changing the "rowIndex" variable will result in some unexpected behaviour of other GREL functions.

Reply:
> My suggestion is that we change the "rowIndex" and the index used by the cross() function both to start from 1, so they will be consistent with the UI index. But I am not sure if changing the "rowIndex" variable will result in some unexpected behaviour of other GREL functions. What's your opinion on this?

@afkbrb: rowIndex is 0-bound everywhere, except in the UI in the preview and the data-grid. I think it would be less breaking and complicated to just change the UI display, and make it 0-bound! ;-)

I'm joking, but I'm also serious. It would be more coherent.

Regards,
   Antoine

Reply:
This has been reviewed and approved. We're not making breaking changes to things. Programmers and computers expect 0-based. Humans expect 1-based. Thus we have two different domains.